monitoring/nethealth.go

```diff
@@ -0,0 +1,321 @@
+	}
+	if c.SeriesCapacity < 0 {
+		errors = append(errors, trace.BadParameter("timeout series capacity cannot be < 0"))
+	}
+	if c.SeriesCapacity == 0 {
+		c.SeriesCapacity = defaultSeriesCapacity
+	}
+	if c.KubeConfig == (KubeConfig{}) {
+		errors = append(errors, trace.BadParameter("kubernetes access config must be provided"))
+	}
+	return trace.NewAggregate(errors...)
+}
+
+// nethealthChecker checks network communication between peers.
+type nethealthChecker struct {
+	// NethealthConfig contains caller specified nethealth checker configuration
+	// values.
+	NethealthConfig
+	// lock access to timeoutStats
+	sync.Mutex
+	// timeoutStats maps a peer to a time series data interval containing the
+	// number of echo timeouts received by the specific peer for the set interval.
+	timeoutStats *holster.TTLMap
+}
+
+// NewNethealthChecker returns a new nethealth checker.
+func NewNethealthChecker(config NethealthConfig) (*nethealthChecker, error) {
+	if err := config.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &nethealthChecker{
+		timeoutStats:    holster.NewTTLMap(timeoutStatsCapacity),
+		NethealthConfig: config,
+	}, nil
+}
+
+// Name returns this checker name
+// Implements health.Checker
+func (c *nethealthChecker) Name() string {
+	return nethealthCheckerID
+}
+
+// Check verifies the network is healthy.
+// Implements health.Checker
+func (c *nethealthChecker) Check(ctx context.Context, reporter health.Reporter) {
+	if err := c.check(ctx); err != nil {
+		reporter.Add(NewProbeFromErr(c.Name(), "failed nethealth test", trace.Wrap(err)))
+		return
+	}
+	reporter.Add(NewSuccessProbe(c.Name()))
+}
+
+func (c *nethealthChecker) check(ctx context.Context) error {
+	metrics, err := c.getMetrics(ctx)
+	if err != nil {
+		return trace.Wrap(err, "failed to get nethealth metrics")
+	}
+
+	updated, err := c.updateStats(metrics)
+	if err != nil {
+		return trace.Wrap(err, "failed to update nethealth timeout stats")
+	}
+
+	return c.verifyNethealth(updated)
+}
+
+// getMetrics returns the network metrics from the local nethealth pod.
+func (c *nethealthChecker) getMetrics(ctx context.Context) ([]*dto.Metric, error) {
+	addr, err := c.getNethealthAddr()
+	if err != nil {
+		return nil, trace.Wrap(err, "failed to get local nethealth address")
+	}
+
+	b, err := fetchMetrics(ctx, addr)
+	if err != nil {
+		log.WithError(err).Warnf("Failed to fetch metrics from %s", addr)
+		return nil, trace.Wrap(err, "failed to fetch metrics")
+	}
+
+	metrics, err := parseMetrics(bytes.NewReader(b))
+	if err != nil {
+		return nil, trace.Wrap(err, "failed to parse metrics")
+	}
+
+	return metrics, nil
+}
+
+// updateStats updates the timeout data with new data points collected in
+// the provided metrics. Returns a list containing the updated keys.
+func (c *nethealthChecker) updateStats(metrics []*dto.Metric) (updated []string, err error) {
+	var errors []error
+	for _, metric := range metrics {
```

Comment:
So what are "metrics" here? Is it a single datapoint per peer representing whether the last nethealth check was success/failure? Or are these multiple binary (1/0) datapoints for each peer? Or number of timeouts for each peer? Could you describe the data model somewhere in the comments please.